### PR TITLE
MBS-10559 maven central publication

### PR DESCRIPTION
**Pull request by dsvoronin**
## Working maven central publication

see `convention.sonatype` and updated release documentation for details

## Reorganized and added some logic in publish conventions

- some hardcoded params goes to extensions
- break some weird dependencies, for example bintray convention sets basic group and version, and also creates teamcity notify task
- expressed actual gradle plugin restriction as preconditions
- add some fail fast logic
- disable test-fixtures.jar publications

## Not in this PR

- Will delete bintray plugin and convention after successfull maven central release
- Removal of jcenter/bintray repositories for dependencies in separate PR

### CODE REVIEW

File: build-logic/publication/src/main/kotlin/convention.publish-base.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            developer {
+                id.set("eugene-krivobokov")
+                name.set("Eugene Krivobokov")
+                url.set("https://github.com/eugene-krivobokov")
+            }
+            developer {
+                id.set("sboishtyan")
+                name.set("Sergey Boishtyan")
+                url.set("https://github.com/sboishtyan")
+            }
```
Comment (eugene-krivobokov) [inline build-logic/publication/src/main/kotlin/convention.publish-base.gradle.kts:43]: Should we add here "service" accounts? Or is it just meta info?

File: build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts [hunk 2]
```diff
@@ -1,8 +1,42 @@
...
-    id("convention.publish-kotlin-library")
+    id("convention.publish-kotlin-base")
+    id("convention.publish-release")
+    id("convention.publish-artifactory")
     id("java-gradle-plugin")
 }
 
 gradlePlugin {
+    // we publish plugins as simple libraries for now,
+    // someone should dive into plugin publication specifics
```
Comment (eugene-krivobokov) [inline build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts:10]: Please, file an issue

File: build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts [hunk 3]
```diff
@@ -1,8 +1,42 @@
...
+                    val pluginName = pluginDeclaration.id.substringAfter("$pluginPrefix.")
+
+                    require(project.name == pluginName) {
+                        // see isAutomatedPublishing note
+                        "Plugin name should be equal to gradle module name to simplify resolving"
+                    }
+
+                    artifactId = pluginName
+                }
+            }
```
Comment (eugene-krivobokov) [inline build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts:40]: Do we need a check for uniqueness plugin ids? Will the same ids be a problem?

File: build-logic/publication/src/main/kotlin/convention.publish-base.gradle.kts (hunk 1 above)
Comment (dsvoronin) [reply build-logic/publication/src/main/kotlin/convention.publish-base.gradle.kts:43]: yes, its metadata

File: build-logic/publication/src/main/kotlin/convention.teamcity.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+plugins {
+    id("convention.publish-base")
+}
+
+/**
+ * https://www.jetbrains.com/help/teamcity/build-script-interaction-with-teamcity.html#BuildScriptInteractionwithTeamCity-ReportingBuildNumber
```
Comment (eugene-krivobokov) [inline build-logic/publication/src/main/kotlin/convention.teamcity.gradle.kts]:
```suggestion
 * https://www.jetbrains.com/help/teamcity/service-messages.html#Reporting+Build+Number
```
?

File: build-logic/testing-convention/src/main/kotlin/convention.test-fixtures.gradle.kts [hunk 5]
```diff
@@ -16,3 +14,7 @@ configure<KotlinJvmProjectExtension> {
     target.compilations.getByName("testFixtures")
         .associateWith(target.compilations.getByName("main"))
 }
+
+val javaComponent = components["java"] as AdhocComponentWithVariants
+javaComponent.withVariantsFromConfiguration(configurations["testFixturesApiElements"]) { skip() }
+javaComponent.withVariantsFromConfiguration(configurations["testFixturesRuntimeElements"]) { skip() }
```
Comment (eugene-krivobokov) [inline build-logic/testing-convention/src/main/kotlin/convention.test-fixtures.gradle.kts]: How about `configurations.matching { name.startsWith("testFixtures") }` to find by prefix?

File: docs/content/internals/Release.md [hunk 6]
```diff
@@ -18,42 +18,41 @@ If it is `Failed` you could release from previous `Succeed` commits or fix probl
 This branch must be persistent. It is used for automation.
 1. Manually run [Integration build](http://links.k.avito.ru/ZA) on the `release branch`.
 1. Manually run [Github publish configuration](http://links.k.avito.ru/releaseAvitoTools) on the `release branch`.
+1. Verify and close staging repo on [nexus](https://oss.sonatype.org/#stagingRepositories). See `Maven central publishing reference article` below
```
Comment (eugene-krivobokov) [inline docs/content/internals/Release.md]:
```suggestion
1. Verify and close staging repo on [nexus](https://oss.sonatype.org/#stagingRepositories). See [Maven central publishing reference article](https://getstream.io/blog/publishing-libraries-to-mavencentral-2021/)
```

Comment (eugene-krivobokov) [inline docs/content/internals/Release.md]: This is a long article with lots of details. 
As next steps, It'd good to extract exact steps in a "How to" manner.

**Review by eugene-krivobokov — APPROVED** [review]

File: build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts (hunk 2 above)
Comment (dsvoronin) [reply build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts:10]: MBS-10660

File: build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts (hunk 3 above)
Comment (dsvoronin) [reply build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts:40]: Can't imagine that it ever be a problem
Don't want premature optimization here

File: build-logic/testing-convention/src/main/kotlin/convention.test-fixtures.gradle.kts (hunk 5 above)
Comment (dsvoronin) [reply build-logic/testing-convention/src/main/kotlin/convention.test-fixtures.gradle.kts]: `Variant for configuration testFixturesAnnotationProcessor does not exist in component java`

don't know how to associate cofigurations which is created by test-fixtures plugin and those that belongs to java software component

added some comments

File: build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts (hunk 3 above)
Comment (eugene-krivobokov) [reply build-logic/publication/src/main/kotlin/convention.publish-gradle-plugin.gradle.kts:40]: Ok

**Pull request merged**